common/migration_mechanism.py

```diff
@@ -0,0 +1,192 @@
+
+
+def _save_schema_version(data_root: str, schema_version: int) -> None:
+    metadata = dict()
+    if os.path.exists(os.path.join(data_root, 'metadata.json')):
+        with open(os.path.join(data_root, 'metadata.json'), 'rt') as f:
+            metadata = json.load(f)
+    metadata['schema_version'] = schema_version
+    with open(os.path.join(data_root, 'metadata.json'), 'wt') as f:
+        json.dump(metadata, f, indent=4, sort_keys=True)
+
+
+def _get_players_to_migrate(data_root: str) -> List[str]:
+    """
+    Get the list of players currently registered
+
+    :param data_root: the root of the data store directory
+    :return: the list of registered player login names
+    """
+    if not os.path.exists(os.path.join(data_root, 'accountdatabase.json')):
+        # No account database
+        return []
+    with open(os.path.join(data_root, 'accountdatabase.json'), 'rt') as f:
+        accounts = json.load(f)
+    return [acc['login_name'] for acc in accounts if 'login_name' in acc]
+
+
+def _get_datastores_for_player(data_root: str, player_name: str) -> List[str]:
+    return glob.glob(os.path.join(data_root, 'players', '%s_*.json' % player_name), recursive=True)
+
+
+def _load_datastores(data_root: str, player_name: str) -> Dict[str, Dict]:
+    # Find datastores for the player
+    player_datastores = _get_datastores_for_player(data_root, player_name)
+    result = OrderedDict()
+    # Load each of them under the top-level dict
+    for ds in player_datastores:
+        datastore_name: str = os.path.splitext(os.path.basename(ds))[0].split('_')[-1]
+        with open(ds, 'rt') as f:
+            result[datastore_name] = json.load(f)
+    return result
+
+
+def _save_datastores(data_root: str, player_name: str, player_data: Dict[str, Dict]) -> None:
+    for ds_name, ds_data in player_data.items():
+        file_path = os.path.join(data_root, 'players', '%s_%s.json' % (player_name, ds_name))
+        with open(file_path, 'wt') as f:
+            json.dump(ds_data, f, indent=4, sort_keys=True)
+
+
+# Migrates all files of all schemas
+def run_migrations(data_root_path: str) -> None:
+    """
+    Run data migrations on all necessary files
+
+    This does no error checking; if any migration fails a ValueError will be raised at that point
+
+    :param data_root_path: the root path of data files to migrate
+    :return: None
+    """
+    existing_version = _load_schema_version(data_root_path)
+
+    # Determine the highest available migration
+    upgraded_version = existing_version
+    while upgraded_version + 1 in _registered_migrations:
+        upgraded_version += 1
+
+    # Exit early if no migration is needed
+    if upgraded_version == existing_version:
+        return
+
+    # Back up all datastores
+    _perform_backups(data_root_path, existing_version)
+
+    # Perform each migration in turn
+    for i in range(existing_version + 1, upgraded_version + 1):
+        _registered_migrations[i]()
```

Comment:
Let's pass the path to the data root to the migrations

Reply:
Done. This also removes the need for passing the data root to the @upgrades_all_players decorator